source/grid/grid_tools.cc

```diff
@@ -4352,284 +4351,102 @@ namespace GridTools
-                if (!box_cell[i].second->is_artificial())
-                  cell_candidate_idx = i;
-
-                // If the cell is not locally owned or at boundary
-                // we check for exceptions
-                if (cell_candidate_idx != -1 &&
-                    (!box_cell[i].second->is_locally_owned() ||
-                     box_cell[i].second->at_boundary()))
-                  use_try = true;
-
-                // If a cell candidate was found we can stop
-                if (cell_candidate_idx != -1)
-                  break;
-              }
-          }
+    // Check all points within a given pair of box and cell
+    const auto check_all_points_within_box =
+      [&](const decltype(*b_tree.begin()) &leaf) {
+        const auto &box       = leaf.first;
+        const auto &cell_hint = leaf.second;
 
-        if (cell_candidate_idx == -1)
-          {
-            // No candidate cell, but the cell might
-            // still be inside the mesh, this is our final check:
-            try
-              {
-                my_pair =
-                  GridTools::find_active_cell_around_point(cache, points[p]);
-                // If we arrive here the cell was not among
-                // the candidates returned by the tree, so we're adding it
-                // by hand
-                cell_candidate_idx = box_cell.size();
-                box_cell.push_back(
-                  std::make_pair(my_pair.first->bounding_box(), my_pair.first));
-              }
-            catch (const GridTools::ExcPointNotFound<dim> &)
-              {
-                missing_points_out.emplace_back(p);
-                continue;
-              }
-          }
-        else
+        for (const auto id :
+             p_tree | bgi::adaptors::queried(!bgi::satisfies(already_found) &&
+                                             bgi::intersects(box)))
           {
-            // We have a candidate cell
-            if (use_try)
+            const auto cell_and_ref =
+              GridTools::find_active_cell_around_point(cache,
+                                                       points[id],
+                                                       cell_hint);
+            const auto &cell      = cell_and_ref.first;
+            const auto &ref_point = cell_and_ref.second;
+
+            if (cell.state() == IteratorState::valid)
               {
-                try
-                  {
-                    my_pair = GridTools::find_active_cell_around_point(
-                      cache, points[p], box_cell[cell_candidate_idx].second);
-                  }
-                catch (const GridTools::ExcPointNotFound<dim> &)
-                  {
-                    missing_points_out.push_back(p);
-                    continue;
-                  }
+                const auto cell_id = get_local_cell_index(cell);
+                qpoints_out[cell_id].emplace_back(ref_point);
+                maps_out[cell_id].emplace_back(id);
```

Comment:
I think a simpler design would be to all three of these operations in the lambda. You'd have to pass the `ref_point` and `id` to that function as well, but at least you don't do the resizing in one place and the emplacement of values in another.